src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_wrap(napi_env e, napi_value jsObject, void* nativeObj,
-                      napi_destruct destructor, napi_weakref* handle) {
+napi_status napi_wrap(napi_env e,
+                      napi_value jsObject,
+                      void* nativeObj,
+                      napi_finalize finalize_cb,
+                      napi_ref* result) {
   NAPI_PREAMBLE(e);
+  CHECK_ARG(jsObject);
+  CHECK_ARG(result);
 
   v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Local<v8::Context> context = isolate->GetCurrentContext();
-  v8::Local<v8::Object> obj;
-  CHECK_TO_OBJECT(context, obj, jsObject);
 
-  v8impl::ObjectWrapWrapper* wrap =
-      new v8impl::ObjectWrapWrapper(obj, nativeObj, destructor);
+  v8::Local<v8::Object> obj = v8impl::V8LocalValueFromJsValue(jsObject).As<v8::Object>();
+  assert(obj->InternalFieldCount() > 0);
+  obj->SetAlignedPointerInInternalField(0, nativeObj);
 
-  if (handle)
-  {
-    return napi_create_weakref(
-      e,
-      v8impl::JsValueFromV8LocalValue(wrap->handle()),
-      handle);
-  }
+  v8impl::Reference* reference = new v8impl::Reference(
```

Comment:
Since we're handing this reference out, user code will have to napi_delete_reference it. If caller doesn't need a handle, we could mark this one to delete itself like the Reference created for external objects. Effectively make the result optional again. Code in leveldown does actually pass nullptr for the handle, though this is probably easy to adapt.

Also, calling finalize_cb is a best effort, right? If we return this ref and calling code immediately deletes it, the persistent will be reset and callback will never be called, as I understand this. If the callback is called, then we'll be working on free'd memory which seems worse. Should we make an effort to call the callback in ~Reference() in this case? Or does persistent reset consistently call the finalizer?